source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.h

```diff
@@ -81,6 +82,36 @@ class GrpcClientImpl : public Client,
 
 using GrpcClientImplPtr = std::unique_ptr<GrpcClientImpl>;
 
+// The client cache for RawAsyncClient for Google grpc so channel is not created for each request.
+class AsyncClientCache : public Singleton::Instance {
+public:
+  AsyncClientCache(Grpc::AsyncClientManager& async_client_manager, Stats::Scope& scope,
+                   ThreadLocal::SlotAllocator& tls)
+      : async_client_manager_(async_client_manager), scope_(scope), tls_slot_(tls.allocateSlot()) {
+    tls_slot_->set([](Event::Dispatcher&) { return std::make_shared<ThreadLocalCache>(); });
+  }
+
+  const Grpc::RawAsyncClientSharedPtr getOrCreateAsyncClient(
+      const envoy::extensions::filters::http::ext_authz::v3::ExtAuthz& proto_config);
+
+private:
+  /**
+   * Per-thread cache.
+   */
+  struct ThreadLocalCache : public ThreadLocal::ThreadLocalObject {
+    ThreadLocalCache() = default;
+    // The client cache stored with key as hash of
+    // envoy::config::core::v3::GrpcService::GoogleGrpc config.
+    absl::flat_hash_map<std::size_t, Grpc::RawAsyncClientSharedPtr> async_clients_;
```

Comment:
Won't this leak RawAsyncClientSharedPtr instances? Each time the config is updated we'll switch to a new key but the old  entries will remain, unused.

Reply:
Yes, Harvey and I have talked about the resource leak when we come up with this solution. But currently this is the best solutions to solve the channel recreation for each request. I'll add a TODO comment to annotate this.

Reply:
I see, I'll need to explore more in the AsyncClientManagerImpl

Reply:
> Have you considered putting this caching in the AsyncClientManagerImpl? You could introduce either a new method there or else a flag on the existing `factoryForGrpcService`. I think that would make this change reusable for other filters and remove the need for your singleton.
> 
> In any event, I think there should be either a config option or a runtime flag to revert to the current behavior.

In order to create a cacheable version of GoogleAsyncClientFactoryImpl and cache the factory in AsyncClientManagerImpl, we need to change the signature of `RawAsyncClientPtr create()` function to return a shared pointer instead of a unique pointer, this potentially break customized code.
So I still prefer current solutions.

Reply:
@htuch for more insight on Google gRPC

Reply:
@zuercher we're planning on doing a clean generic solution along the lines you suggest this quarter. I think it's fine to land this as-is, since it repeats a pattern we're fairly comfortable with elsewhere in the code base.

Reply:
Ok. Can we put a TODO here to track?

Reply:
Done, and have issue https://github.com/envoyproxy/envoy/issues/13417 tracking this